Tests/PostgresNIOTests/New/Extensions/PSQLBackendMessageEncoder.swift

```diff
@@ -0,0 +1,279 @@
+    func encode(data message: PSQLBackendMessage, out buffer: inout ByteBuffer) throws {
+        switch message {
+        case .authentication(let authentication):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(authentication, into: &buffer)
+            
+        case .backendKeyData(let keyData):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(keyData, into: &buffer)
+            
+        case .bindComplete,
+             .closeComplete,
+             .emptyQueryResponse,
+             .noData,
+             .parseComplete,
+             .portalSuspended:
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(EmptyPayload(), into: &buffer)
+            
+        case .commandComplete(let string):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(StringPayload(string), into: &buffer)
+            
+        case .dataRow(let row):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(row, into: &buffer)
+            
+        case .error(let errorResponse):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(errorResponse, into: &buffer)
+            
+        case .notice(let noticeResponse):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(noticeResponse, into: &buffer)
+            
+        case .notification(let notificationResponse):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(notificationResponse, into: &buffer)
+            
+        case .parameterDescription(let description):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(description, into: &buffer)
+            
+        case .parameterStatus(let status):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(status, into: &buffer)
+            
+        case .readyForQuery(let transactionState):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(transactionState, into: &buffer)
+            
+        case .rowDescription(let description):
+            buffer.writeBackendMessageID(message.id)
+            self.encodePayload(description, into: &buffer)
+            
+        case .sslSupported:
+            buffer.writeInteger(UInt8(ascii: "S"))
+            
+        case .sslUnsupported:
+            buffer.writeInteger(UInt8(ascii: "N"))
+        }
+    }
+    
+    private struct EmptyPayload: PSQLMessagePayloadEncodable {
+        func encode(into buffer: inout ByteBuffer) {}
+    }
+    
+    private struct StringPayload: PSQLMessagePayloadEncodable {
+        var string: String
+        init(_ string: String) { self.string = string }
+        func encode(into buffer: inout ByteBuffer) {
+            buffer.writeNullTerminatedString(self.string)
```

Comment:
Okay, it's not your fault, but this is reminding me of a peeve that's been annoying the <censored> out of me for some time. Can we:
1) Move the multiple copies of various `ByteBuffer` utilities scattered around PostgresNIO, MySQLNIO, etc. into something they all use, like AsyncKit.
2) Fix the `writeFoo()` utilities to work the same way as `ByteBuffer`'s builtins - specifically in returning the number of bytes written, and preferably being based on `setFoo()` variants of themselves as well.
3) Having done those, no longer have to do things like `buffer.writerIndex - startIndex`.
4) Does this actually need to be an `encode()`-style method/protocol? Why can't it be something like `func write(into buffer: inout ByteBuffer) -> Int`? No generic-ness is lost, and there's no pretense of `Codable`-like semantics in a place where they have no real business being anyway.

(Again, there's nothing actually _wrong_ with the code you've written - it's just still trapped in some of the crummy code smells Tanner left lying all over the place 😰.)

Reply:
I think we should keep `PostgresNIO` dependency free except for the NIO and Crypto dependencies.